### DISCUSSION TITLE
Create custom TextBlock with special drawing

### DISCUSSION THREAD

**Discussion by JouriM2** [AvaloniaUI/Avalonia#15879] [Q&A] [answered]
I need to display my data, which is simple text string, but I need to highlight some characters in this string with special background.
These text strings are indentifiers and I need to visually display if user used impossible characters in them.
The task itself is very simple and at start I think what I need to create custom control based on TextBlock and add little highlighting like AccessText do.
But Render method is sealed and all methods I need to get character positions are private\internal.
How to complete my task without fully implementing TextBlock again?

PS: I need a control\method to display text like this:
S\<O>MEN\<A>ME
where characters in <> must be highlighted with custom background color.

PPS: These texts are displayed in cells of grid with hundreds of lines, so to combine several separated TextBlocks for each cell is highly inefficient and slow.

**Comment by stevemonaco** [comment, marked as answer]
The right way is probably to work with inlines. You can split your text into several `Run` and apply the background only to parts that have the highlighted text. You'll need some balance between number of `TextBlock`s. If each `TextBlock` is hundreds of lines of text, you'll probably have a bad time, but it wasn't clear to me if that's the case per-`TextBlock` or just the entire page.

You can check the [Inlines docs](https://docs.avaloniaui.net/docs/reference/controls/detailed-reference/textblock#inlines) for a brief example of how to use `Run`. You would be building an `InlineCollection` of `Run`s via code somewhere and assigning to `TextBlock.Inlines`.

The most performant is probably a bespoke implementation as `TextBlock` is "general use".

**Reply by JouriM2** [reply]
Thanks for the answer.

To be sure Im get all right.
In Avalonia I  need to use existing controls and infrastructure exactly as they are and, if need to extend some of them, I need to reimplement control itself and all infrastructure it uses from the ground.
Avalonia made to make as hard as possible for me to reuse implementation details of existing controls.
Am I right?

For me its very surprising and discouraging method to work with _universal_ GUI framework.
All frameworks I work before (in C++ and pascal) are targeted to maximum reuse and object oriented.
Even ones which is not written on object oriented languages allow users to reuse its part.
I was thinking what this is the main goal of all frameworks...
It seems Avalonia is not the case and  this tool is not for me.
Thank you.

**Reply by timunie** [reply]
Also AvaloniaEdit may be what you are looking for. 

Otherwise if Avalonia isn't your cup of tea, that's fine. Thanks for your feedback anyways.

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/AvaloniaEdit

**Reply by stevemonaco** [reply]
> To be sure Im get all right. In Avalonia I need to use existing controls and infrastructure exactly as they are and, if need to extend some of them, I need to reimplement control itself and all infrastructure it uses from the ground. Avalonia made to make as hard as possible for me to reuse implementation details of existing controls. Am I right?
> 
> For me its very surprising and discouraging method to work with _universal_ GUI framework. All frameworks I work before (in C++ and pascal) are targeted to maximum reuse and object oriented. Even ones which is not written on object oriented languages allow users to reuse its part. I was thinking what this is the main goal of all frameworks... It seems Avalonia is not the case and this tool is not for me. Thank you.

You're making some large, inaccurate assumptions here. There are tradeoffs in any design. An IDE quality text control has huge overhead when you're trying to display simple text labels. Likewise, the Avalonia `TextBlock` is not good for huge amounts of text that gets modified/virtualized because it doesn't use text ropes and it caches the entire text vector geometry. Different tools for different jobs. As I said, Inlines will work for you, but you're going to pay some cost. Most times, this is ok. Sometimes, it's not.

Most of the underlying components that `TextBlock` relies upon can be reused. It also doesn't really matter if `Render` is sealed or not as that isn't what does the work. `RenderTextLayout` does and you can override it. eg. https://github.com/stevemonaco/AvaloniaDemos/tree/master/TextOutlineAnimation which derives from `TextBlock` and adds text outlining capabilities.

**Reply by JouriM2** [reply]
Thanks alot!
I missed what RenderTextLayout is protected.
Now my small tool is complete and finally display data as requested by users.

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/96559149-c3dc-4edc-acfc-9e938d99e8d6)

Using little extension of existing class:

```C#
public class CellControl : TextBlock {
    public IBrush? CmpColor;

    public CellControl( CellData f ) {
        VerticalAlignment = Avalonia.Layout.VerticalAlignment.Center;

        Text = f.Text;
        Background = null;

        switch( f.Cmp ) {
            case CompareRC.Equal:
                CmpColor = null;
                break;
            case CompareRC.Diff:
                CmpColor = new ImmutableSolidColorBrush( MainWindow.ColorDiff );
                break;
            case CompareRC.New:
                CmpColor = new ImmutableSolidColorBrush( MainWindow.ColorNew );
                break;
            case CompareRC.Delete:
                CmpColor = new ImmutableSolidColorBrush( MainWindow.ColorDelete );
                break;
        }
    }
    protected override void RenderTextLayout( DrawingContext context, Point origin ) {
        if( CmpColor != null )
            context.FillRectangle( CmpColor, new Rect( Bounds.Size ) );

        var txt = Text;
        if( MainWindow.IsRusHilight && txt is not null ) {
            for( int n = 0; n < txt.Length; n++ )
                if( Rus.IsRusChar( txt[ n ] ) ) {
                    var rect = TextLayout!.HitTestTextPosition( n );
                    context.FillRectangle( MainWindow.BrushRus, rect );
                }
        }
        base.RenderTextLayout( context, origin );
    }
}
```

As you can see above I dont need "huge amount of text", just simple line and TextBlock is simplest control I found for this.
Are where any simpler controls for this?

**Reply by stevemonaco** [reply]
`TextBlock` is the simplest `Control` to render text. Anything simpler requires implementation "from scratch", but `TextLayout` is public and does most of the heavy lifting. Of course, you can use SkiaSharp to draw onto a `SKCanvas`, too. The normal case is straightforward, but I've never done highlighting.

The overall approach looks good. You might want to inflate rectangles on the same line to send less `FillRectangle` calls. Looks cleaner than building `Inlines` anyways!

**Reply by JouriM2** [reply]
> You might want to inflate rectangles on the same line to send less FillRectangle calls.

Right, in "production app" im definately do this but this tool doesnt worth it.

Just for information.
I was looking for reasons why Render()-s are made sealed but found nothing.
In some other wrameworks, like f.i. Uno, it make sence since it uses native controls in all platforms and render parts are very different on them.
But as far as I can see Avalonia allways uses offscreen renderer (Skia) to draw content.
What the point to disable access to render methonds?

**Reply by stevemonaco** [reply]
AFAIK, they were sealed to prevent breaking changes in the future if Avalonia changes how primitives (ie. self-drawing controls) work. See https://github.com/AvaloniaUI/Avalonia/pull/10299 and there are very few primitives. Most controls are `TemplatedControl` or `UserControl` (in apps). These can still have their `Render` overridden, but it's really uncommon.

In general, Avalonia 11 heavily restricted public API to limit breaking changes going forward.

There's no guarantee that Skia will stay the main renderer in the long-term, so these primitives are ultimately drawn using Avalonia's backend rendering abstraction: either SkiaSharp or Direct2D (mostly exists to mitigate Avalonia's design becoming too coupled to Skia).

**Reply by JouriM2** [reply]
I saw this thread but where no answer to "why" where.
For me its very strange way to decoupling from specific API.
In any frameworks I know (from platform oriented like MFC/OWL to universal like VCL in delphi or Swing in java), if some universality needed framework give special abstraction to user and makes it persistent for all implementations used.
It may be done like in java: to separate control representation and display its representation (very complex way to use) or like in VCL (easy way), where abstract class "Context" exists and implement it on any platform (some of them are WinAPI, Skia, OpenGL in android, etc).
But in any framework I never saw method of "privating" things, making impossible to extend unique controls, to guard for "possible" future changes.
Anyway, thanks for the tip above.
